### ISSUE TITLE
Issue in block_cache::wait_io()'s error handling

### ISSUE THREAD

**Issue by mingnus** [jthornber/thin-provisioning-tools#68]
Hi Joe,

There's an issue in block_cache::wait_io()'s error handling. libaio defines io_event::res as unsigned long, but block_cache::wait_io() thought it was signed, thus the condition "(e.res < 0)" never establishes.

Kernel defines io_event::res as signed 64:
(include/uapi/linux/aio_abi.h)
```
/* read() from /dev/aio returns these structures. */
struct io_event {
        __u64           data;           /* the data field from the iocb */
        __u64           obj;            /* what iocb this event came from */
        __s64           res;            /* result code for this event */
        __s64           res2;           /* secondary result */
};
```
But libaio defines io_event::res as unsigned long:
https://git.fedorahosted.org/cgit/libaio.git/tree/src/libaio.h
```
struct io_event {
	PADDEDptr(void *data, __pad1);
	PADDEDptr(struct iocb *obj,  __pad2);
	PADDEDul(res,  __pad3);
	PADDEDul(res2, __pad4);
};
```

We might need to apply this patch. I'm not an expert in aio and filesystem, so I'm not sure whether the value of io_event::res would be less than INT_MIN or not.
(As I know, in Direct IO, io_event::res might be the value returned by file_operations::fsync(), so it was int originally)

```
diff --git a/block-cache/block_cache.cc b/block-cache/block_cache.cc
index e7e6f14..cb29bb5 100644
--- a/block-cache/block_cache.cc
+++ b/block-cache/block_cache.cc
@@ -213,7 +213,7 @@ block_cache::wait_io()
                if (e.res == block_size_ << SECTOR_SHIFT)
                        complete_io(*b, 0);
 
-               else if (e.res < 0)
+               else if (static_cast<int>(e.res) < 0)
                        complete_io(*b, e.res);
 
                else {
```

**Comment by jthornber** [comment]
Good catch, send me a pull request.

On another note are you familiar at all with the Rust language?  I've been
playing with it recently and am wondering about moving all tools
development over to it.

**Comment by mingnus** [comment]
Commit sent.

Do you mean you are considering to rewrite it in Rust?
I'm afraid I cannot help currently. I can only program in C, C++, Java, and a little Python.
But I'll keep following your work.

**Comment by mingnus** [comment]
Solved by pull-request #69. Validated by using a faulty md RAID and a dm-error device.
